assets/js/modules/analytics/common/timzezone-select.js

```diff
@@ -0,0 +1,142 @@
+/**
+ * TimezoneSelect component.
+ *
+ * Site Kit by Google, Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { useState, useMemo } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import {
+	Select,
+	Option,
+} from '../../../material-components';
+import { countries } from './countries';
+import classnames from 'classnames';
+
+const { default: { country: allCountries } } = countries;
+
+const TimezoneSelect = ( { timezone, setTimezone, hasError } ) => {
+	const [ selectedCountry, setSelectedCountry ] = useState( timezone );
+	const [ selectedTimezoneID, setSelectedTimezoneID ] = useState( '' );
+	let multiTimezone,
+		selectedTimezoneDisplay = selectedTimezoneID,
+		foundTimezone = false;
+	const getTimezoneSelector = () => {
+		const response = (
+			<div >
+				<span>
+					<Select
+						className={ classnames(
+							'googlesitekit-analytics__select-timezone first',
+							{ 'mdc-text-field--error': hasError }
+						) }
+						name="country"
+						enhanced
+						value={ selectedCountry }
+						onEnhancedChange={ ( i, item ) => {
+							setTimezone( item.dataset.value );
+							setSelectedCountry( item.dataset.value );
+							setSelectedTimezoneID( item.dataset.value );
+						} }
+						label={ __( 'Country', 'google-site-kit' ) }
+						outlined
+					>
+						{
+							allCountries && allCountries
+								.map( ( aCountry ) => {
+									// If the selected timezone is in this country, the country should be selected.
+									let value = aCountry.defaultTimeZoneId;
+									const timezoneMatch = aCountry.timeZone.find( ( tz ) => tz.timeZoneId === timezone );
+									if ( timezoneMatch ) {
+										foundTimezone = true;
+										value = timezoneMatch.timeZoneId;
+										if ( aCountry.timeZone.length > 1 ) {
+											multiTimezone = aCountry.timeZone;
+										} else {
+											setSelectedTimezoneID( aCountry.timeZone[ 0 ].displayName );
+											selectedTimezoneDisplay = aCountry.timeZone[ 0 ].displayName;
+											multiTimezone = false;
+										}
+										setSelectedCountry( timezoneMatch.timeZoneId );
+									}
+
+									return (
+										<Option
+											key={ aCountry.displayName }
+											value={ value }
+										>
+											{ aCountry.displayName }
+										</Option>
+									);
```

Comment:
This function should only map countries into `Option`s, but there are `set` calls being made in the iteration.

This is likely rather unperformant (hence the need for `useMemo`) as you're setting state multiple times, every time this is rendered.

The `set*` calls should only be invoked by user-action (i.e. `onEnhancedChange`) or `useEffect` with specific dependencies as well as a guard - likely only to set if not already set.

Reply:
Good point. The set calls are about setting the selection status when we find a matching timezone in the data. I will break that part into a separate function and only call it once on initialization.

Reply:
@aaemnnosttv I tried moving the search out of the display logic, and this does remove the set state calls, however I found I still need the caching to make the dropdown performant. See https://github.com/google/site-kit-wp/pull/1389/commits/d639a51a9b3f2cada4f255ddfc4b0a6e4d18e002 and let me know what you think.

Part of the issue here is that this is a two part dropdown, the second part depends on the first part (when you change the country, you may or may not get a second dropdown with timezones for that country).

Adding useMemo ensures the the drop-downs only re-render when timezone changes and seems reasonable, however I am open to any optimizations you can find please feel free to push updates if you have ideas.

Reply:
I managed to simplify a bit further by only calling `timezoneInCountries` once when we first load the component. I run this once using a flag `timezoneChecked` which is set to true after the first run. The main point of this function is to see if the timezone set in WordPress can be matched to an actual timezone and to fall back to the browser timezone if not.

Documented better in https://github.com/google/site-kit-wp/pull/1389/commits/552bd9cf0174d32d7f4c3fe0a9795dc065a83012.